File: depth-first-search/DFS.java
import java.util.*;

public class DFS {
    public static void dfs(Map<Integer, List<Integer>> graph, int start, Set<Integer> visited) {
        visited.add(start);
        System.out.print(start + " ");
        
        List<Integer> neighbors = graph.getOrDefault(start, new ArrayList<>());
        for (int neighbor : neighbors) {
            if (!visited.contains(neighbor)) {
                dfs(graph, neighbor, visited);
            }
        }
    }
    
    public static void main(String[] args) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        graph.put(0, Arrays.asList(1, 2));
        graph.put(1, Arrays.asList(2));
        graph.put(2, Arrays.asList(0, 3));
        graph.put(3, Arrays.asList(3));
        
        Set<Integer> visited = new HashSet<>();
        dfs(graph, 2, visited);
    }
}

File: depth-first-search/DFS.kt
fun dfs(graph: Map<Int, List<Int>>, start: Int, visited: MutableSet<Int>) {
    visited.add(start)
    print("$start ")
    
    graph[start]?.forEach { neighbor ->
        if (neighbor !in visited) {
            dfs(graph, neighbor, visited)
        }
    }
}

fun main() {
    val graph = mapOf(
        0 to listOf(1, 2),
        1 to listOf(2),
        2 to listOf(0, 3),
        3 to listOf(3)
    )
    
    val visited = mutableSetOf<Int>()
    dfs(graph, 2, visited)
}